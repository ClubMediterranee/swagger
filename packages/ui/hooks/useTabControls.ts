import {
  Dispatch,
  RefObject,
  SetStateAction,
  useCallback,
  useEffect,
  useRef,
  useState,
} from 'react';

import { KEYS } from './constants';

export const useTabControls = (
  items: Array<object>,
  initialIndex: number,
  callback?: (arg0: number) => void,
): [
  number,
  Dispatch<SetStateAction<number>>,
  RefObject<HTMLButtonElement[]>,
  () => void,
  () => void,
] => {
  const [activeItem, setActiveItem] = useState(initialIndex);
  const refsArray = useRef<HTMLButtonElement[]>([]);

  const goTo = useCallback(
    (direction: -1 | 1) => {
      const newIndex = (activeItem + direction + items.length) % items.length;
      setActiveItem(newIndex);
      refsArray.current[newIndex] && refsArray.current[newIndex].focus();
    },
    [activeItem, items.length],
  );

  const goToNext = useCallback(() => {
    const isRTL = document.documentElement.dir === 'rtl';
    isRTL ? goTo(-1) : goTo(1);
  }, [goTo]);

  const goToPrev = useCallback(() => {
    const isRTL = document.documentElement.dir === 'rtl';
    isRTL ? goTo(1) : goTo(-1);
  }, [goTo]);

  const onKeyDown = useCallback(
    (event: KeyboardEvent) => {
      if (items.length > 1) {
        switch (event.which) {
          case KEYS.ARROW_RIGHT:
          case KEYS.ARROW_DOWN:
            event.preventDefault();
            goToNext();
            return;
          case KEYS.ARROW_LEFT:
          case KEYS.ARROW_UP:
            event.preventDefault();
            goToPrev();
            return;
          case KEYS.HOME:
            event.preventDefault();
            setActiveItem(0);
            return;
          case KEYS.END:
            event.preventDefault();
            setActiveItem(items.length - 1);
            return;
          default:
        }
      }
    },
    [goToNext, goToPrev, items.length],
  );

  useEffect(() => {
    const refs = refsArray.current;

    refs.forEach((ref) => ref.addEventListener('keydown', onKeyDown, false));

    return () => {
      refs.forEach((ref) => ref?.removeEventListener('keydown', onKeyDown, false));
    };
  }, [items, onKeyDown]);

  useEffect(() => {
    callback && callback(activeItem);
  }, [activeItem, callback]);

  return [activeItem, setActiveItem, refsArray, goToNext, goToPrev];
};
