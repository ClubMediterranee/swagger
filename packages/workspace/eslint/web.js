module.exports = {
  root: true,
  env: {
    browser: true,
    es2021: true
  },
  extends: [
    "react-app",
    "react-app/jest",
    "plugin:jsx-a11y/strict",
    "plugin:prettier/recommended",
    "plugin:cypress/recommended"
  ],
  plugins: ["prettier", "react-hooks"],
  rules: {
    "@typescript-eslint/no-unused-vars": "error",
    "prettier/prettier": "error",
    "react-hooks/rules-of-hooks": "error", // Checks rules of Hooks
    "react-hooks/exhaustive-deps": "warn", // Checks effect dependencies
    "cypress/assertion-before-screenshot": "warn",
    "cypress/no-force": "warn",
    "cypress/no-pause": "error",
    "testing-library/no-unnecessary-act": "warn",
    "testing-library/prefer-screen-queries": "warn",
    "testing-library/prefer-presence-queries": "warn",
    "testing-library/no-wait-for-multiple-assertions": "warn",
    "testing-library/no-node-access": ["warn", { allowContainerFirstChild: true }],
    "testing-library/no-render-in-setup": "warn"
  }
};
