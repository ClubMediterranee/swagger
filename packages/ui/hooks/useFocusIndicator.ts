import { SpringValue, useSpring } from '@react-spring/web';
import { RefObject, useEffect, useRef } from 'react';

import useResizeObserver from './useResizeObserver';

const useFocusIndicator = (
  activeItem: number,
  itemsRefs: RefObject<HTMLButtonElement[]>,
  spacerRef?: RefObject<HTMLDivElement>,
  horizontalGrowth = 0,
): [
  {
    [key: string]: SpringValue<string> | SpringValue<number>;
  },
  RefObject<HTMLDivElement>,
  SpringValue<number>,
] => {
  const container = useRef<HTMLDivElement>(null);
  const { width: containerWidth } = useResizeObserver<HTMLDivElement>({ ref: container });
  const [focusIndicatorStyle, focusIndicatorApi] = useSpring(() => ({
    from: {
      left: 0,
      height: 0,
      width: 0,
      translateX: 0,
      opacity: 0,
    },
  }));

  const [{ scrollLeft }, scrollApi] = useSpring(() => ({
    from: { scrollLeft: 0 },
  }));

  useEffect(() => {
    if (!itemsRefs.current || !container.current) {
      return;
    }

    const spacerWidth = spacerRef?.current?.offsetWidth || 0;
    const rect = itemsRefs.current[activeItem].getBoundingClientRect();
    const scrollLeft = container.current?.scrollLeft;

    focusIndicatorApi.start({
      to: {
        height: rect.height,
        width: rect.width + horizontalGrowth * 2,
        translateX:
          rect.left +
          scrollLeft -
          container.current.getBoundingClientRect().left -
          horizontalGrowth,
        opacity: 1,
      },
      config: { mass: 1, tension: 300, friction: 30 },
    });
    scrollApi.start({
      to: {
        scrollLeft: scrollLeft + rect.left - spacerWidth,
      },
    });
  }, [
    activeItem,
    containerWidth,
    focusIndicatorApi,
    horizontalGrowth,
    itemsRefs,
    scrollApi,
    spacerRef,
  ]);

  return [focusIndicatorStyle, container, scrollLeft];
};

export default useFocusIndicator;
