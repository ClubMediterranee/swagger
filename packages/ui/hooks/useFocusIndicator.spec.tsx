import { animated } from '@react-spring/web';
import { fireEvent, render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import useFocusIndicator from './useFocusIndicator';
import { useTabControls } from './useTabControls';

describe('useFocusIndicator', () => {
  const TestComponent = () => {
    const items = [{ label: 'test1' }, { label: 'test2' }];
    const [activeItem, setActiveItem, refsArray] = useTabControls(items, 0);

    const [focusIndicatorStyle, scrollArea, scrollLeft] = useFocusIndicator(activeItem, refsArray);

    return (
      <div className="relative flex" data-testid="scrollarea" ref={scrollArea}>
        <animated.div
          className="pointer-events-none absolute"
          style={{
            ...focusIndicatorStyle,
          }}
          data-testid="focusindicator"
        />
        {items.map((item, index) => (
          <button
            key={index}
            ref={(element) => {
              return refsArray.current && (refsArray.current[index] = element as HTMLButtonElement);
            }}
            onClick={() => {
              setActiveItem(index);
            }}
            style={{ width: 100 }}
          >
            {item.label}
          </button>
        ))}
        <animated.div data-testid="scrollleft">{scrollLeft.to((x) => x)}</animated.div>
      </div>
    );
  };

  it('should update focus indicator position when active item changes', () => {
    const { getByText, getByTestId } = render(<TestComponent />);

    const scrollArea = getByTestId('scrollarea');
    const button2 = getByText('test2');
    const focusIndicator = getByTestId('focusindicator');

    //take a snapshot of the initial state
    expect(focusIndicator).toMatchSnapshot('initial state');

    userEvent.click(button2);
    fireEvent.mouseMove(scrollArea, { clientX: 200, clientY: 0 });
    fireEvent.mouseMove(scrollArea, { clientX: -200, clientY: 0 });

    //take a snapshot of the state after clicking the second button
    expect(focusIndicator).toMatchSnapshot('after clicking second button');
  });

  it("shouldn't work without essential refs", () => {
    // this is purely for code coverage, to test the early returns in useFocusIndicator
    const Component = () => {
      const items = [{ label: 'test1' }, { label: 'test2' }];
      const [activeItem] = useTabControls(items, 0);

      const props = useFocusIndicator(activeItem, {
        current: null,
      });

      return <div>{props.toString()}</div>;
    };

    render(<Component />);
    expect('nothing').toBe('nothing');
  });
});
