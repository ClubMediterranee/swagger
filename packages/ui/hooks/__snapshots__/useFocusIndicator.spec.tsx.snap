// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useFocusIndicator should update focus indicator position when active item changes: after clicking second button 1`] = `
<div
  class="pointer-events-none absolute"
  data-testid="focusindicator"
  style="left: 0px; height: 0px; width: 0px; opacity: 0; transform: none;"
/>
`;

exports[`useFocusIndicator should update focus indicator position when active item changes: initial state 1`] = `
<div
  class="pointer-events-none absolute"
  data-testid="focusindicator"
  style="left: 0px; height: 0px; width: 0px; opacity: 0; transform: none;"
/>
`;
